File: Ahorn/entities/timeCrystal.jl
module TimeCrystal

using ..Ahorn, Maple

@mapdef Entity "vitellary/timecrystal" Crystal(x::Integer, y::Integer,
	oneUse::Bool=false, stopLength::Number=2.0, respawnTime::Number=2.5,
	untilDash::Bool=false, immediate::Bool=false, entityTypesToIgnore::String="",
    timeScale::Number=0.0)

const placements = Ahorn.PlacementDict(
    "Time Crystal (Crystalline)" => Ahorn.EntityPlacement(
        Crystal,
		"point"
    ),
	"Time Crystal (Until Dash) (Crystalline)" => Ahorn.EntityPlacement(
        Crystal,
		"point",
		Dict{String, Any}(
			"untilDash" => true
        )
    )
)

Ahorn.editingOrder(entity::Crystal) = String["x", "y", "stopLength", "respawnTime", "timeScale",
    "entityTypesToIgnore", "oneUse", "immediate", "untilDash"]

function Ahorn.selection(entity::Crystal)
    x, y = Ahorn.position(entity)

    return Ahorn.getSpriteRectangle("objects/crystals/time/idle00.png", x, y, jx=0.5, jy=0.5)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::Crystal)
	untilDash = get(entity.data, "untilDash", false)
	if untilDash
		Ahorn.drawSprite(ctx, "objects/crystals/time/untildash/idle00.png", 0, 0, jx=0.5, jy=0.5)
	else
		Ahorn.drawSprite(ctx, "objects/crystals/time/idle00.png", 0, 0, jx=0.5, jy=0.5)
	end
end

end

File: Loenn/entities/time_crystal.lua
local timeCrystal = {}

timeCrystal.name = "vitellary/timecrystal"
timeCrystal.depth = -100

timeCrystal.fieldInformation = {
    stopLength = {
        minimumValue = 0.0
    },
    respawnTime = {
        minimumValue = 0.0
    },
    timeScale = {
        minimumValue = 0.0
    }
}

timeCrystal.placements = {
    {
        name = "normal",
        data = {
            oneUse = false,
            stopLength = 2.0,
            respawnTime = 2.5,
            untilDash = false,
            immediate = false,
            entityTypesToIgnore = "",
            timeScale = 0.0
        }
    },
    {
        name = "until_dash",
        data = {
            oneUse = false,
            stopLength = 2.0,
            respawnTime = 2.5,
            untilDash = true,
            immediate = false,
            entityTypesToIgnore = "",
            timeScale = 0.0
        }
    }
}

function timeCrystal.texture(room, entity)
    return entity.untilDash and "objects/crystals/time/untildash/idle00" or "objects/crystals/time/idle00"
end

return timeCrystal
